### PR TITLE
Add KRICO experiment

### PR DESCRIPTION
**Pull request by Creatone**
Add KRICO experiment to SWAN ( www.krico.gda.pl )

Summary of changes:
- Added science workload.

Testing done:
-

### CODE REVIEW

File: pkg/executor/openstack.go [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+		return nil, err
+	}
+
+	floatingIP, err := stack.findFloatingIP()
+	if err != nil {
+		log.Error(err.Error())
+		return nil, err
+	}
+
+	image, err := images.IDFromName(stack.client, stack.config.Image)
```
Comment (iwankgb) [inline pkg/executor/openstack.go]: You **must** check this error.

File: pkg/executor/openstack.go [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+
+	instanceName := fmt.Sprintf("krico.%s", uuid.New())
+
```
Comment (iwankgb) [inline pkg/executor/openstack.go]: Is there any particular reason for adding an empty line here?

File: pkg/executor/openstack.go [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+	aggregate, err := aggregates.Get(stack.client, hostAggregateIDFlag.Value()).Extract()
+	if err != nil {
+		err = errors.Wrapf(err, "%s Couldn't get host aggregate info", executorLogPrefix)
+		log.Error(err.Error())
+		return nil, err
+	}
+
+	stack.config.HostAggregate.Name = aggregate.Name
+	stack.config.HostAggregate.AvailabilityZone = aggregate.AvailabilityZone
+
```
Comment (iwankgb) [inline pkg/executor/openstack.go]: Remove empty line, please (I do not understand why it is there).

File: pkg/executor/openstack.go [hunk 4]
```diff
@@ -0,0 +1,707 @@
...
+	log.Infof("%s Created remote executor", executorLogPrefix)
+
+	remoteHandler, err := remote.Execute(command)
+	if err != nil {
+		log.Errorf("%s Couldn't execute %q on %s (%s)", executorLogPrefix, command, instanceName, floatingIP)
+		return nil, err
+	}
+
+	exitCode, err := remoteHandler.ExitCode()
```
Comment (iwankgb) [inline pkg/executor/openstack.go]: You should not expect exit code to be available while command is still running.

File: pkg/executor/openstack.go [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+	stdoutFile.Close()
+	stderrFile.Close()
+
+	taskHandle := &OpenstackTaskHandle{
+		command:        command,
+		hostIP:         floatingIP,
+		stdoutFilePath: stdoutFileName,
+		stderrFilePath: stderrFileName,
+		instance:       instance.ID,
+		os:             &stack,
```
Comment (iwankgb) [inline pkg/executor/openstack.go]: Why do you need to pass instance of executor into the task?

File: pkg/executor/openstack.go [hunk 6]
```diff
@@ -0,0 +1,707 @@
...
+		return nil
+	}
+
+	log.Debugf("%s delete instance %q", taskHandleLogPrefix, th.instance)
+
+	th.requestStop <- struct{}{}
+
+	log.Debugf("%s waiting for instance %q to stop", taskHandleLogPrefix, th.instance)
+
+	<-th.stopped
```
Comment (iwankgb) [inline pkg/executor/openstack.go]: Use a timeout channel, just in case.

File: pkg/executor/openstack.go [hunk 7]
```diff
@@ -0,0 +1,707 @@
...
+
+	log.Debugf("%s instance %q stop", taskHandleLogPrefix, th.instance)
+
+	th.requestDelete <- struct{}{}
+
+	<-th.deleted
```
Comment (iwankgb) [inline pkg/executor/openstack.go]: Use a timeout channel, just in case.

File: pkg/executor/openstack.go [hunk 8]
```diff
@@ -0,0 +1,707 @@
...
+				*watcher.running = false
+			case "RESCUED":
+				*watcher.running = true
+			case "STOPPED":
+				*watcher.running = false
+			case "SOFT_DELETED":
+				*watcher.running = false
+				watcher.requestDelete <- struct{}{}
+			}
+			time.Sleep(time.Second)
```
Comment (iwankgb) [inline pkg/executor/openstack.go]: Make sleep time configurable, please.

File: pkg/snap/sessions/krico/krico.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krico
+
+import (
+	"github.com/intelsdi-x/swan/pkg/executor"
+	"github.com/intelsdi-x/swan/pkg/snap"
+	"time"
+	//"github.com/intelsdi-x/snap/scheduler/wmap"
```
Comment (iwankgb) [inline pkg/snap/sessions/krico/krico.go]: Delete commented out lines, please.

File: experiments/krico/krico-classification/main.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	errutil.CheckWithContext(err, "Cannot send request to KRICO for importing monitor samples!")
+	log.Infof("Monitor samples imported!")
+
+	for _, instance := range instances {
+		predictedCategory, err := krico.Classify(context.Background(), &api.ClassifyRequest{InstanceId: instance})
+		errutil.CheckWithContext(err, fmt.Sprintf("Cannot send request to KRICO for classify %v instance.", instance))
+		log.Infof("Predicted category for %v instance: %q", instance, predictedCategory.ClassifiedAs)
+	}
+
+	defer conn.Close()
```
Comment (iwankgb) [inline experiments/krico/krico-classification/main.go]: You must call `defer conn.Close()` as soon as you determine that `conn` is a valid object.

File: experiments/krico/workloads/caching.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+	// Wait until Snap will load plugins
+	// time.Sleep(time.Second*10)
+
+	//
+	//	KRICO parameters
+	//
+
+	// Calculate workload parameters
+	memory := float64(memcachedConfig.MaxMemoryMB / 1024) // total cache size [GiB]
+	ratio, err := strconv.ParseFloat(mutilateConfig.Update, 64) // estimated get vs set ratio [0.0 - 1.0]
```
[the remark this answers is not in the document]
Comment (iwankgb) [reply experiments/krico/workloads/caching.go]: This is still valid, I believe.

File: experiments/krico/workloads/utils.go [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+const (
+	// TypeCaching is constant name of caching workload.
+	TypeCaching = "caching"
+)
+
+var (
+	snapStartCommand = "systemctl restart snap-telemetry"
+	aggressorAddress = conf.NewStringFlag("aggressor_address", "IP address of aggressor node.", "127.0.0.0")
```
Comment (iwankgb) [inline experiments/krico/workloads/utils.go]: You do not launch any agressors.

File: experiments/krico/workloads/caching.go [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+	//	Prepare configuration.
+	loadGeneratorConfig := mutilate.DefaultMutilateConfig()
+	loadGeneratorConfig.MemcachedHost = workloadHandle.Address()
+	loadDuration := sensitivity.LoadDurationFlag.Value()
+	maxQPS := sensitivity.PeakLoadFlag.Value()
+
+	//	Prepare executor.
+	loadGeneratorExecutorConfig := executor.DefaultRemoteConfig()
+	loadGeneratorExecutor, err := executor.NewRemote(aggressorAddress.Value(), loadGeneratorExecutorConfig)
```
Comment (iwankgb) [inline experiments/krico/workloads/caching.go]: You do not launch any agressors.

File: experiments/krico/workloads/caching.go [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+
+	//	Prepare configuration.
+	loadGeneratorConfig := ycsb.DefaultYcsbConfig()
+	loadGeneratorConfig.RedisHost = workloadHandle.Address()
+	loadDuration := sensitivity.LoadDurationFlag.Value()
+	maxQPS := sensitivity.PeakLoadFlag.Value()
+
+	//	Prepare executor.
+	loadGeneratorExecutorConfig := executor.DefaultRemoteConfig()
+	loadGeneratorExecutor, err := executor.NewRemote(aggressorAddress.Value(), loadGeneratorExecutorConfig)
```
Comment (iwankgb) [inline experiments/krico/workloads/caching.go]: You do not launch any agressors.

File: experiments/krico/workloads/caching.go [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+	//	Start load on workload.
+	loadGeneratorHandle, err := loadGeneratorLauncher.Load(maxQPS, loadDuration)
+	errutil.CheckWithContext(err, "Cannot start Mutilate task!")
+
+	//	In the end stop load on workload.
+	defer loadGeneratorHandle.Stop()
+
+	//	Wait until load generating finishes.
+	loadGeneratorHandle.Wait(0)
```
Comment (iwankgb) [inline experiments/krico/workloads/caching.go]: Do not ignore errors, please.

File: experiments/krico/workloads/caching.go [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+	//	Start load on workload.
+	loadGeneratorHandle, err := loadGeneratorLauncher.Load(maxQPS, loadDuration)
+	errutil.CheckWithContext(err, "Cannot start YCSB Redis task !")
+
+	//	In the end stop load on workload.
+	defer loadGeneratorHandle.Stop()
+
+	//	Wait until load generating finishes.
+	loadGeneratorHandle.Wait(0)
```
Comment (iwankgb) [inline experiments/krico/workloads/caching.go]: Do not ignore errors, please.

File: experiments/krico/workloads/utils.go [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+	TypeCaching = "caching"
+	TypeOltp = "oltp"
+	TypeScience = "science"
+	TypeStreaming = "streaming"
+	TypeWebserving = "webserving"
+)
+
+var (
+	experimentID string
+	snapStartCommand = "service snap-telemetry restart"
```
[the remark this answers is not in the document]
Comment (iwankgb) [reply experiments/krico/workloads/utils.go]: Still a issue, I think.

File: experiments/krico/workloads/utils.go (hunk 12 above)
Comment (iwankgb) [reply experiments/krico/workloads/utils.go]: This flag must be renamed and moved to appropriate file: the one where it is used.

File: experiments/krico/workloads/utils.go [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+func GetInstanceCgroup(hypervisorInstanceName string, hypervisorAddress string) (string, error){
+	conn, err := libvirt.NewConnect("qemu+ssh://root@"+hypervisorAddress+"/system")
+	if err != nil {
+		return "", fmt.Errorf("couldn't connect to libvirt: %v", err)
+	}
+	defer conn.Close()
+
+	domain, err := conn.LookupDomainByName(hypervisorInstanceName)
+	if err != nil {
+		return "", fmt.Errorf("couldn't get instance domain: %v", err)
```
[the remark this answers is not in the document]
Comment (iwankgb) [reply experiments/krico/workloads/utils.go]: Still valid.

File: experiments/krico/workloads/utils.go [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+func Initialize(experimentId string) {
+	experimentID = experimentId
+}
+
+func GetInstanceCgroup(hypervisorInstanceName string, hypervisorAddress string) (string, error){
+	conn, err := libvirt.NewConnect("qemu+ssh://root@"+hypervisorAddress+"/system")
+	if err != nil {
+		return "", fmt.Errorf("couldn't connect to libvirt: %v", err)
```
[the remark this answers is not in the document]
Comment (iwankgb) [reply experiments/krico/workloads/utils.go]: Still valid.

File: experiments/krico/workloads/utils.go [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+	if err != nil {
+		return "", fmt.Errorf("couldn't get instance domain: %v", err)
+	}
+
+	instanceID, err := domain.GetID()
+	if err != nil {
+		return "", fmt.Errorf("couldn't get instance domain id: %v", err)
+	}
+
+	instanceName := strings.Replace(hypervisorInstanceName, "-", `\x2d`, -1)
```
Comment (iwankgb) [inline experiments/krico/workloads/utils.go]: Replace magic string with a constant.

File: experiments/krico/workloads/utils.go [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+
+	cgroup := "machine.slice:machine-qemu" + `\x2d` + fmt.Sprint(instanceID) + `\x2d` + instanceName + ".scope"
```
Comment (iwankgb) [inline experiments/krico/workloads/utils.go]: Ditto.

File: glide.yaml [hunk 22]
```diff
@@ -32,6 +32,10 @@ import:
   version: master
   subpackages:
   - ssh
+- package: golang.org/x/net
+  version: master
+- package: golang.org/x/text
+  version: master
```
[the remark this answers is not in the document]
Comment (iwankgb) [reply glide.yaml]: Can you answer, please?

File: pkg/executor/openstack.go [hunk 23]
```diff
@@ -0,0 +1,707 @@
...
+type OpenstackAuthConfig struct {
+	Username   string
+	Password   string
+	TenantID   string
+	DomainName string
+	Endpoint   string
+}
+
+type hostAggregate struct {
+	Name             string
```
[the remark this answers is not in the document]
Comment (iwankgb) [reply pkg/executor/openstack.go]: Still valid.

File: pkg/executor/openstack.go [hunk 24]
```diff
@@ -0,0 +1,707 @@
...
+// OpenstackConfig defines OpenStack instance configuration data.
+type OpenstackConfig struct {
+	Auth          gophercloud.AuthOptions
+	Flavor        OpenstackFlavor
+	Image         string
+	User          string
+	SSHKeyPath    string
+	Name          string
+	ID            string
+	Hypervisor    hypervisor
```
[the remark this answers is not in the document]
Comment (iwankgb) [reply pkg/executor/openstack.go]: Still valid.

File: experiments/krico/workloads/utils.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (Creatone) [reply experiments/krico/workloads/utils.go]: There is no code like this

[the remark this answers is not in the document]
Comment (iwankgb) [reply experiments/krico/workloads/utils.go]: You are right indeed.

File: glide.yaml (hunk 22 above)
[the remark this answers is not in the document]
Comment (Creatone) [reply glide.yaml]: "golang.org/x/net" - used by generated code from protobuf
second one, not used

File: experiments/krico/workloads/utils.go (hunk 12 above)
Comment (Creatone) [reply experiments/krico/workloads/utils.go]: But this flag would be used in many workloads that's why it is in utils

Comment (iwankgb) [reply experiments/krico/workloads/utils.go]: In Swan "aggressor" is a synthetic workload that competes with high priority application for a resource. This is wrong place and wrong name, I believe.

File: glide.yaml (hunk 22 above)
[the remark this answers is not in the document]
Comment (iwankgb) [reply glide.yaml]: Remove unused dependency, please.

File: experiments/krico/workloads/utils.go (hunk 20 above)
Comment (Creatone) [reply experiments/krico/workloads/utils.go]: Could you said what you mean, more precisely?

File: experiments/krico/README.md [hunk 25]
```diff
@@ -0,0 +1,196 @@
...
+## Running the experiments
+
+#### Metric gathering
+```bash
+krico-metric-gathering
+```
+Parameters available:
+
+```
+-aggressor_address string
```
Comment (iwankgb) [inline experiments/krico/README.md]: Fix name of the flag.

File: experiments/krico/README.md [hunk 26]
```diff
@@ -0,0 +1,196 @@
...
+```
+
+#### Classification
+```bash
+krico-classification
+```
+Parameters available:
+
+```
+-aggressor_address string
```
Comment (iwankgb) [inline experiments/krico/README.md]: Fix name of the flag.

File: experiments/krico/krico-classification/main.go [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+	errutil.CheckWithContext(err, "Cannot send request to KRICO for importing monitor samples!")
+	log.Infof("Monitor samples imported!")
+
+	for _, instance := range instances {
+		predictedCategory, err := krico.Classify(context.Background(), &api.ClassifyRequest{InstanceId: instance})
+		errutil.CheckWithContext(err, fmt.Sprintf("Cannot send request to KRICO for classify %v instance.", instance))
+		log.Infof("Predicted category for %v instance: %q", instance, predictedCategory.ClassifiedAs)
+	}
+
+	err = conn.Close()
```
Comment (iwankgb) [inline experiments/krico/krico-classification/main.go]: This call *must* be deferred and moved to line 81.

File: experiments/krico/krico-metric-gathering/main.go [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+	_, err = krico.ImportMetricsFromSwanExperiment(context.Background(), &api.ImportMetricsFromSwanExperimentRequest{ExperimentId: experimentID})
+	errutil.CheckWithContext(err, "Cannot send request to KRICO for importing metrics!")
+
+	_, err = krico.RefreshClassifier(context.Background(), &api.RefreshClassifierRequest{})
+	errutil.CheckWithContext(err, "Cannot send request to KRICO for refreshing classifier!")
+
+	_, err = krico.RefreshPredictor(context.Background(), &api.RefreshPredictorRequest{})
+	errutil.CheckWithContext(err, "Cannot send request to KRICO for refreshing predictor!")
+
+	err = conn.Close()
```
Comment (iwankgb) [inline experiments/krico/krico-metric-gathering/main.go]: This call must be deferred and moved to line 80.

File: experiments/krico/krico-prediction/main.go [hunk 29]
```diff
@@ -0,0 +1,145 @@
...
+	prediction, err := krico.Predict(context.Background(), &api.PredictRequest{
+		Category:   workloadCategory.Value(),
+		Image:      workloadImage.Value(),
+		Parameters: parameters,
+	})
+	errutil.CheckWithContext(err, fmt.Sprintf("Cannot make prediction for %q workload!", workloadCategory.Value()))
+
+	log.Infof("Prediction for %v workload: \nRequirements: %q \nFlavor: %q \nHost aggregate: %q", workloadCategory.Value(), prediction.Requirements, prediction.Flavors, prediction.HostAggregates)
+
+	err = conn.Close()
```
Comment (iwankgb) [inline experiments/krico/krico-prediction/main.go]: This call must be deferred and moved to line 85.

File: experiments/krico/workloads/caching.go (hunk 15 above)
Comment (iwankgb) [reply experiments/krico/workloads/caching.go]: Error is still being ignored.

File: experiments/krico/workloads/caching.go (hunk 16 above)
Comment (iwankgb) [reply experiments/krico/workloads/caching.go]: Error is still being ignored.

File: glide.yaml (hunk 22 above)
[the remark this answers is not in the document]
Comment (iwankgb) [reply glide.yaml]: If ``golang.org/x/text`` is unused then it must be removed from ``glide.yaml``.

**Pull request merged**